pennylane/fourier/advanced_spectrum.py

```diff
@@ -0,0 +1,364 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains a transform that computes the frequency spectrum of a quantum
+circuit including classical preprocessing within the QNode."""
+from itertools import product
+from functools import wraps
+from collections import OrderedDict
+from inspect import signature
+import numpy as np
+import pennylane as qml
+from .utils import _get_spectrum, _join_spectra
+
+
+def _process_ids(encoding_args, argnum, qnode):
+    r"""Process the passed ``encoding_args`` and ``argnum`` or infer them from
+    the QNode signature.
+
+    Args:
+        encoding_args (dict[str, list[tuple]] or set): Parameter index dictionary;
+            keys are argument names, values are index tuples for that argument
+            or an ``Ellipsis``. If a ``set``, all values are set to ``Ellipsis``
```

Comment:
What is an `Ellipsis`?

Reply:
I learned about it in this PR! It has an alias, namely `...` and is used for example in NumPy to indicate a flexible number of axes (example taken from [here](https://stackoverflow.com/questions/118370/how-do-you-use-the-ellipsis-slicing-syntax-in-python)):
```python
>>> from numpy import arange
>>> a = arange(16).reshape(2,2,2,2)
>>> a[..., 0].flatten()
array([ 0,  2,  4,  6,  8, 10, 12, 14])
```
Although it is not super well-known I feel like this is a perfect use case because it is meant to say "fill in all indices that make sense for this argument".